salt/state.py

```diff
@@ -1261,6 +1294,15 @@ def compile_high_data(self, high):
                                   not isinstance(val, six.string_types)):
                                 # Invalid name, fall back to ID
                                 chunk[key] = name
+                            elif ('__sls__' in body and
+                                  '__env__' in body and
+                                  key in STATE_REQUISITE_KEYWORDS):
+                                _val = type(val)()
```

Comment:
Not sure if there's a better pattern for this, but I didn't want to change the type of whatever was being passed through here. `type(val)()`, from my understanding, will be either `dict` or `OrderedDict`, both of which have vararg constructors, so this works. This technique doesn't seem to be used anywhere else, so I'd be interested in suggestions for a more idiomatic approach.